src/policy/policy.cpp

```diff
@@ -154,6 +154,123 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
+                if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SIGVERSION_BASE))
+                    return TX_BAD_P2SH;
+                if (stack.empty())
+                    return TX_BAD_P2SH;
+                prevScript = CScript(stack.back().begin(), stack.back().end());
+            }
+
+            int witnessversion = 0;
+            std::vector<unsigned char> witnessprogram;
+
+            // Non-witness program must not be associated with any witness
+            if (!prevScript.IsWitnessProgram(witnessversion, witnessprogram) && !tx.wit.vtxinwit[i].IsNull())
+                return TX_BAD_WITNESS;
+
+            // Witness for P2WPKH must have 2 stack items. Pubkey must match the witness program.
+            // Signature must be <= 73 bytes (BIP66). Since signature size is variable, it is not impossible to bloat a
+            // 70-byte signature to 73-byte and trigger fee rejection for marginal cases but the impact is negligible.
+            if (witnessversion == 0 && witnessprogram.size() == 20) {
+                if (tx.wit.vtxinwit[i].scriptWitness.stack.size() != 2)
+                    return TX_BAD_WITNESS;
+                std::vector<unsigned char> hashPubKey(20);
+                std::vector<unsigned char> pubKey = tx.wit.vtxinwit[i].scriptWitness.stack[1];
+                if (tx.wit.vtxinwit[i].scriptWitness.stack[0].size() > 73)
+                    return TX_BAD_WITNESS;
+                CHash160().Write(begin_ptr(pubKey), pubKey.size()).Finalize(begin_ptr(hashPubKey));
+                if (hashPubKey != witnessprogram)
+                    return TX_BAD_WITNESS;
+            }
+
+            // P2WSH requires at least 1 witness stack item but we already know the witness is not empty.
+            // Make sure the witnessScript size is <= 10000 bytes (MAX_SCRIPT_SIZE) and matches witness program.
+            // Max standard witnessScript size is 3600 bytes
+            if (witnessversion == 0 && witnessprogram.size() == 32) {
+                CScript witnessScript = CScript(tx.wit.vtxinwit[i].scriptWitness.stack.back().begin(), tx.wit.vtxinwit[i].scriptWitness.stack.back().end());
+                uint256 hashWitnessScript;
+                if (witnessScript.size() > MAX_SCRIPT_SIZE)
+                    return TX_BAD_WITNESS;
+                if (witnessScript.size() > MAX_STANDARD_P2WSH_SCRIPT_SIZE)
+                    ret = TX_NONSTD_BIG_P2WSH;
+                CSHA256().Write(begin_ptr(witnessScript), witnessScript.size()).Finalize(hashWitnessScript.begin());
+                if (memcmp(hashWitnessScript.begin(), &witnessprogram[0], 32))
+                    return TX_BAD_WITNESS;
+
+                /*
+                 * Count the witness stack size without the witnessScript.
+                 *
+                 * Max standard witness stack size is 100.
+                 * Due to the implicit CLEANSTACK rule and nOpCount limit of 201, the maximum consensus valid
```

Comment:
Could you spell out the math here? I'm not a Script wizard and can't figure what's being asserted here.